src/libraries/Common/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -273,38 +306,59 @@ public override int KeySize
                 base.KeySize = value;
 
                 ThrowIfDisposed();
-                _key.Dispose();
-                _key = new ECOpenSsl(this);
+
+                FreeKey();
+                _key = new Lazy<SafeEvpPKeyHandle>(GenerateKeyFromSize);
             }
         }
 
         public override void GenerateKey(ECCurve curve)
         {
             ThrowIfDisposed();
-            _key.GenerateKey(curve);
+
+            FreeKey();
+            _key = new Lazy<SafeEvpPKeyHandle>(SafeEvpPKeyHandle.GenerateECKey(curve, out int keySize));
 
             // Use ForceSet instead of the property setter to ensure that LegalKeySizes doesn't interfere
             // with the already loaded key.
-            ForceSetKeySize(_key.KeySize);
+            ForceSetKeySize(keySize);
         }
 
         public override void ImportParameters(ECParameters parameters)
         {
             ThrowIfDisposed();
-            _key.ImportParameters(parameters);
-            ForceSetKeySize(_key.KeySize);
+
+            FreeKey();
+            _key = new Lazy<SafeEvpPKeyHandle>(SafeEvpPKeyHandle.GenerateECKey(parameters, out int keySize));
+
+            // Use ForceSet instead of the property setter to ensure that LegalKeySizes doesn't interfere
+            // with the already loaded key.
+            ForceSetKeySize(keySize);
+        }
+
+        private SafeEvpPKeyHandle GenerateKeyFromSize()
+        {
+            return SafeEvpPKeyHandle.GenerateECKey(KeySize);
         }
 
         public override ECParameters ExportExplicitParameters(bool includePrivateParameters)
         {
             ThrowIfDisposed();
-            return ECOpenSsl.ExportExplicitParameters(_key.Value, includePrivateParameters);
+
+            using (SafeEcKeyHandle ecKey = Interop.Crypto.EvpPkeyGetEcKey(_key.Value))
+            {
+                return ECOpenSsl.ExportExplicitParameters(ecKey, includePrivateParameters);
+            }
         }
 
         public override ECParameters ExportParameters(bool includePrivateParameters)
         {
             ThrowIfDisposed();
-            return ECOpenSsl.ExportParameters(_key.Value, includePrivateParameters);
+
+            using (SafeEcKeyHandle ecKey = Interop.Crypto.EvpPkeyGetEcKey(_key.Value))
+            {
+                return ECOpenSsl.ExportParameters(ecKey, includePrivateParameters);
+            }
```

Comment:
This is really chatty at the P/Invoke boundary.  I'd expect us to never talk about `EC_KEY*` anymore except for the `ECDsaOpenSsl(IntPtr)` ctor (and the same for ECDH), and we just update the shim library to do what we used to do for EC_KEY but with EVP_PKEY.

The way this is written:

* Fetch _key.Value to $key
* $key.DangerousAddRef();
* $eckey = new SafeEcKeyHandle();
* P/invoke, $tmp = CryptoNative_EvpPkeyGetEcKey($key.handle);
* $eckey.handle = $tmp;
* $key.DangerousRelease();
* $eckey.DangerousAddRef();
* // export parameters
* $eckey.DangerousRelease();
* $eckey.Dispose(); => another P/Invoke

If the shim is adjusted, it is instead just


* Fetch _key.Value to $key
* $key.DangerousAddRef();
* // export parameters
* $key.DangerousRelease();

All work we already had to do, but with fewer intermediate finalizable objects, and fewer P/invoke crossings.

Reply:
I agree this would be ideal but unless I'm missing something that doesn't look particularly straight forward and there don't seem to be much perf benefit from this since it's not in the hot path anywhere. I'd imagine we want to get rid of all old APIs usages because we want to get rid of OpenSSL 1.x support eventually? I'll open the work item for this unless it's easier than it looks. I assume you mean to use these new APIs: https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_get_params.html ?

Reply:
My meta-point is that if we're working in terms of EVP_PKEY, then we want our ExportParameters API to be based on EVP_PKEY rather than EC_KEY.

Reply:
Opened: https://github.com/dotnet/runtime/issues/105173 to track this work